pkg/cache/informer.go

```diff
@@ -0,0 +1,265 @@
+	informer := runtime.Must(store.GetInformer(ctx, &v1beta1.ApplicationRevision{}))
+	informer.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if rev, ok := obj.(*v1beta1.ApplicationRevision); ok {
+				in.RemapRevision(rev)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if rev, ok := oldObj.(*v1beta1.ApplicationRevision); ok {
+				in.UnmapRevision(rev)
+			}
+			if rev, ok := newObj.(*v1beta1.ApplicationRevision); ok {
+				in.RemapRevision(rev)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if rev, ok := obj.(*v1beta1.ApplicationRevision); ok {
+				in.UnmapRevision(rev)
+			}
+		},
+	})
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			t0 := time.Now()
+			compDefPruned := in.ComponentDefinitionCache.Prune(duration)
+			traitDefPruned := in.TraitDefinitionCache.Prune(duration)
+			wsDefPruned := in.WorkflowStepDefinitionCache.Prune(duration)
+			klog.Infof("DefinitionCache prune finished. ComponentDefinition: %d(-%d), TraitDefinition: %d(-%d), WorkflowStepDefinition: %d(-%d). Time cost: %d ms.",
+				in.ComponentDefinitionCache.Size(), compDefPruned,
+				in.TraitDefinitionCache.Size(), traitDefPruned,
+				in.WorkflowStepDefinitionCache.Size(), wsDefPruned,
+				time.Since(t0).Microseconds())
+			time.Sleep(duration)
+		}
+	}
+}
+
+// DefaultDefinitionCache the default definition cache
+var DefaultDefinitionCache = singleton.NewSingleton(NewDefinitionCache)
+
+func filterUnnecessaryField(o client.Object) {
+	_ = k8s.DeleteAnnotation(o, "kubectl.kubernetes.io/last-applied-configuration")
+	o.SetManagedFields(nil)
+}
+
+func wrapTransformFunc[T client.Object](fn func(T)) kcache.TransformFunc {
+	return func(i interface{}) (interface{}, error) {
+		if o, ok := i.(T); ok {
+			filterUnnecessaryField(o)
+			fn(o)
+			return o, nil
+		}
+		return i, nil
+	}
+}
+
+// AddInformerTransformFuncToCacheOption add informer transform func to cache option
+// This will filter out the unnecessary fields for cached objects and use definition cache
+// to reduce the duplicated storage of same definitions
+func AddInformerTransformFuncToCacheOption(opts *cache.Options) {
+	if OptimizeInformerCache {
+		if opts.TransformByObject == nil {
+			opts.TransformByObject = map[client.Object]kcache.TransformFunc{}
+		}
+		opts.TransformByObject[&v1beta1.ApplicationRevision{}] = wrapTransformFunc(func(rev *v1beta1.ApplicationRevision) {
+			if OptimizeApplicationRevisionDefinitionStorage {
+				DefaultDefinitionCache.Get().RemapRevision(rev)
+			}
+		})
+		opts.TransformByObject[&v1beta1.Application{}] = wrapTransformFunc(func(app *v1beta1.Application) {})
+		opts.TransformByObject[&v1beta1.ResourceTracker{}] = wrapTransformFunc(func(rt *v1beta1.ResourceTracker) {})
+	}
+}
+
+// NewClientDisableCacheFor get ClientDisableCacheFor for building controller
+func NewClientDisableCacheFor() []client.Object {
```

Comment:
```suggestion
func Resources2DisableCache() []client.Object {
```

Reply:
This is aligned with the naming of controller-runtime manager. See https://github.com/kubevela/kubevela/blob/afc5c8c2db4b5a322ad785fbd10ad84bcd9a54e6/cmd/core/app/server.go#L165

Reply:
they're different, the naming is trying to become a sentence:

Client disable cache for "some resources", so the field should be "some resources", not copy-pase the name.

Reply:
ok